x-pack/plugins/fleet/server/services/agents/reassign_action_runner.ts

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { SavedObjectsClientContract, ElasticsearchClient } from '@kbn/core/server';
+
+import type { Agent, BulkActionResult } from '../../types';
+
+import { AgentReassignmentError, HostedAgentPolicyRestrictionRelatedError } from '../../errors';
+
+import { appContextService } from '../app_context';
+
+import { ActionRunner } from './action_runner';
+
+import { errorsToResults, bulkUpdateAgents } from './crud';
+import { createAgentAction } from './actions';
+import { getHostedPolicies, isHostedAgent } from './hosted_agent';
+import { BulkActionTaskType } from './bulk_actions_resolver';
+
+export class ReassignActionRunner extends ActionRunner {
+  protected async processAgents(agents: Agent[]): Promise<{ items: BulkActionResult[] }> {
+    return await reassignBatch(
+      this.soClient,
+      this.esClient,
+      this.actionParams! as any,
+      agents,
+      {},
+      undefined,
+      true
+    );
+  }
+
+  protected getTaskType() {
+    return BulkActionTaskType.REASSIGN_RETRY;
+  }
+
+  protected getActionType() {
+    return 'POLICY_REASSIGN';
+  }
+}
+
+export async function reassignBatch(
+  soClient: SavedObjectsClientContract,
+  esClient: ElasticsearchClient,
+  options: {
+    newAgentPolicyId: string;
+    actionId?: string;
+    total?: number;
+  },
+  givenAgents: Agent[],
+  outgoingErrors: Record<Agent['id'], Error>,
+  agentIds?: string[],
+  skipSuccess?: boolean
+): Promise<{ items: BulkActionResult[] }> {
+  const errors: Record<Agent['id'], Error> = { ...outgoingErrors };
+
+  const hostedPolicies = await getHostedPolicies(soClient, givenAgents);
+
+  // which are allowed to unenroll
+  const agentResults = await Promise.allSettled(
```

Comment:
nit: don't think this needs to be async?

Reply:
updated